Add support to redirect table reads from Hive to Iceberg

findepi opened the pull request:
Hive Connector redirects Iceberg table access to the configured
Iceberg catalog.

This change adds implementation of `HiveTableRedirectionsProvider`,
plugging in Hive->Iceberg redirects and leveraging existing framework.

This is based on work from several authors. In particular, this is based
on https://github.com/trinodb/trino/pull/8340 with conflicts resolved, and retrofit to current APIs in Hive connector.

As tests show, not all the statements support redirects properly and
this needs to be followed up upon.

findepi commented:
per @ssheikin's https://github.com/trinodb/trino/pull/8340#discussion_r762143226, added @MiguelWeezardo as co-author.

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveConfig.java:
```diff
@@ -1092,6 +1094,19 @@ public boolean isLegacyHiveViewTranslation()
+        return icebergCatalogName;
+    }
+
+    @Config("hive.iceberg-catalog-name")
```
nit: if we plan to have more of those maybe it makes sense to group them somehow:
```
hive.redirects.iceberg-catalog-name
```
?

findepi replied:
discussed offline as we decided to keep the name as-is for now

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveWriterFactory.java:
```diff
@@ -256,8 +257,12 @@ public HiveWriterFactory(
+                .map(propertyMetadata -> immutableEntry(
+                        propertyMetadata.getName(),
+                        session.getProperty(propertyMetadata.getName(), propertyMetadata.getJavaType())))
+                // The session properties collected here are used for events only. Filter out nulls to avoid problems with downstream consumers
```
Is there a reason why we do not care about properties which are set to `null` in events?

findepi replied:
I am not aware of any consumer of these events. This must be some extension point for someone.
The only usage i found looks like a dummy consumer, at least this is what the commit message says.
https://github.com/trinodb/trino/blob/cea58426e0a8f4cfe330bc805f88d12ab2d3fc4e/plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveEventClient.java#L31-L38

Including `null` values could be a breaking change to downstream even consumers (e.g. if the `[Immutable]Map.copyOf(event.getSessionProperties())`. Converting values to `Optional` would be a breaking change to. In the absence of more information, i deemed that omitting null values is the least breaking of all options.

losipiuk commented on testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/environment/EnvSinglenodeHiveRedirectionToIceberg.java:
```diff
@@ -0,0 +1,52 @@
+import static org.testcontainers.utility.MountableFile.forHostPath;
+
+@TestsEnvironment
+public class EnvSinglenodeHiveRedirectionToIceberg
```
This probably does not matter here but I heard many times in the past that we should refrain from using singlenode envs as they provide worse test coverage than multinode ones.

findepi replied:
Thanks for reminding me. We have 27 or something singlenode environments. And yes, here it doesn't matter, as tested functionality is really coordinator-only. Let me keep it as is.

losipiuk commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveRedirectionToIceberg.java:
```diff
@@ -0,0 +1,837 @@
+    }
+
+    @Test(groups = {HIVE_REDIRECTION_TO_ICEBERG, PROFILE_SPECIFIC_TESTS}, dataProvider = "schemaAndPartitioning")
+    public void testHiveToIcebergInsert(String schema, boolean partitioned)
```
nit: move above `testIcebergToHiveInsert` (so we always have same flavor of test ordered hive->incberg first icebarg->hive later)

findepi replied:
> so we always have same flavor of test ordered hive->incberg first icebarg->hive later


i tried to make it so, but actually it's not as ordered (initial part of the class uses different ordering than neighbourhood here).

i can reorder. but i also though about splitting the class into two (hive->iceberg redirects in `hive` test package and iceberg->hive redirects in `iceberg` package). wdyt?

losipiuk replied:
Splitting is better idea IMO. If you can structure it so not too much code is repeated

findepi replied:
discussed offline, and decide to split.

losipiuk commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveRedirectionToIceberg.java:
```diff
@@ -0,0 +1,837 @@
+    }
+
+    @Test(groups = {HIVE_REDIRECTION_TO_ICEBERG, PROFILE_SPECIFIC_TESTS})
+    public void testHiveToIcebergDelete()
```
move above `testIcebergToHiveDelete`

findepi replied:
splitting the class instead

losipiuk commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveRedirectionToIceberg.java:
```diff
@@ -0,0 +1,837 @@
+    }
+
+    @Test(groups = {HIVE_REDIRECTION_TO_ICEBERG, PROFILE_SPECIFIC_TESTS})
+    public void testHiveToIcebergUpdate()
```
move above `testIcebergToHiveUpdate`

findepi replied:
splitting the class instead

losipiuk commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveRedirectionToIceberg.java:
```diff
@@ -0,0 +1,837 @@
+    }
+
+    @Test(groups = {HIVE_REDIRECTION_TO_ICEBERG, PROFILE_SPECIFIC_TESTS})
+    public void testHiveToIcebergDropTable()
```
move above `testIcebergToHiveDropTable`

losipiuk replied:
and for others :)

losipiuk commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveRedirectionToIceberg.java:
```diff
@@ -0,0 +1,837 @@
+        onTrino().executeQuery("ALTER TABLE " + hiveTableName + " ADD COLUMN some_new_column double");
+
+        // TODO: ALTER TABLE succeeded, but new column was not added
+        Assertions.assertThat(onTrino().executeQuery("DESCRIBE " + icebergTableName).column(1))
```
can you static import?

findepi replied:
i cannot. assertj's `Assertions.assertThat` conflicts with tempto's.

losipiuk commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveRedirectionToIceberg.java:
```diff
@@ -0,0 +1,837 @@
+
+        onTrino().executeQuery("ALTER TABLE " + hiveTableName + " ADD COLUMN some_new_column double");
+
+        // TODO: ALTER TABLE succeeded, but new column was not added
```
can you create a github issue which lists missing functionalities?

findepi replied:
I don't think it's necessary in this case. https://github.com/trinodb/trino/pull/8340 is going to fix this problems (once https://github.com/trinodb/trino/pull/8340#discussion_r748709426 is resolved).

hashhar reviewed (commented):
Do we want to test SHOW STATS redirection too?

What about a non-default hive.timestamp-precision? Redirects will return different results.
Same for the use_column_name configs in tables which have different schema across partitions.

findepi commented:
> What about a non-default hive.timestamp-precision? Redirects will return different results.

`hive.timestamp-precision` doesn't matter, since there is no Hive table. Hive doesn't know the table schema, so it doesn't know whether there are any timestamps.

findepi commented:
> What about a non-default hive.timestamp-precision? Redirects will return different results.

`hive.timestamp-precision` doesn't matter, since there is no Hive table. Hive doesn't know the table schema, so it doesn't know whether there are any timestamps.

> Same for the use_column_name configs in tables which have different schema across partitions.

same

findepi commented:
> Do we want to test SHOW STATS redirection too?

added

Also, split the test class into two (https://github.com/trinodb/trino/pull/10173#discussion_r762908545).

aierate commented:
Do I have to use  like 
 USE a-catalog.myschema;
 SELECT * FROM mytable;
or
SELECT * FROM a-catalog.myschema.mytable;
?
Instead of using SELECT * FROM myschema.mytable without using a-catalog or USE a-catalog?
